Add builtin implementation for std targets.

Dirbaio opened the pull request:
This adds an `std` Cargo feature that enables a critical-section implementation for `std`.

IMO this is the only impl that we should have "builtin". Rust's `std` is essentially "stable forever" so the impl won't need breaking changes, and it should work soundly on all targets where `std` is supported.

Note this uses Rust 1.63's "const Mutex::new()", so this makes our MSRV effectively that for std targets. I think it should be acceptable.

eldruin reviewed (commented):
Seems fine to me but I am no expert on CS. Maybe @adamgreig could review?
I noticed now that there is no MSRV notice or CI job at all here.
Being such a foundational crate I think we should have one, even if it is 1.63.
It's definitely not a breaking change if there was none before, right? :wink:

Dirbaio commented:
Added 1.63 MSRV notice to the README.

> It's definitely not a breaking change if there was none before, right?

It's not breaking because 1.63 is only actually required if you enable the `std` feature, which didn't exist before.

reitermarkus commented on src/std.rs:
```diff
@@ -0,0 +1,40 @@
+use core::{cell::Cell, mem::MaybeUninit};
```
Any reason not to use `std` here also?

reitermarkus commented on src/std.rs:
```diff
@@ -0,0 +1,40 @@
+                return true;
+            }
+
+            // Not acquied in the current thread, acquire it.
```
```suggestion
            // Not acquired in the current thread, acquire it.
```

reitermarkus commented on src/std.rs:
```diff
@@ -0,0 +1,40 @@
+
+    unsafe fn release(nested_cs: bool) {
+        if !nested_cs {
+            // safety: as per the acquire/release safety contract, release can only be called
```
```suggestion
            // SAFETY: As per the acquire/release safety contract, release can only be called
```

reitermarkus commented on src/std.rs:
```diff
@@ -0,0 +1,40 @@
+    unsafe fn release(nested_cs: bool) {
+        if !nested_cs {
+            // safety: as per the acquire/release safety contract, release can only be called
+            // if the critical section is acquired in the current thread.
```
```suggestion
            // if the critical section is acquired in the current thread,
```

reitermarkus commented on src/std.rs:
```diff
@@ -0,0 +1,40 @@
+
+std::thread_local!(static IS_LOCKED: Cell<bool> = Cell::new(false));
+
+struct CriticalSection;
```
Maybe name this `MutexCriticalSection` instead?

reitermarkus commented on Cargo.toml:
```diff
@@ -14,6 +14,11 @@ categories = [
 
 [features]
 
+# Enable the critical-section implementation for std.
```
```suggestion
# Enable a critical-section implementation based on `std::sync::Mutex`.
```

reitermarkus commented on .github/workflows/clippy.yml:
```diff
@@ -10,5 +10,13 @@ jobs:
       - name: Build
         run: cargo clippy
+      - name: Build for std
+        run: cargo clippy --features std
```
Missing newline here.

reitermarkus reviewed (changes requested):
Some small suggestions, but in general looks good to me.

Dirbaio commented:
Thanks @reitermarkus for the review, I've incorporated the feedback.

eldruin reviewed (commented):
Could you also add this feature to the test CI workflow?

adamgreig commented:
* I don't like the asymmetry in `acquire` being "lock, then store true to cell" while `release` is "unlock, then store false to cell", but I appreciate that since the cell is thread-local and the thread can't pre-empt itself it probably doesn't matter that another thread might have taken the mutex between us unlocking it and us storing false back to the cell. Indeed it means we hold the mutex lock for very slightly less time which is probably beneficial. But, it's more "obvious" to keep the lock until we've cleared the cell. What do you think?
* Otherwise the std impl all seems fine to me :+1: 
* We should document the new feature in the README so that it also appears in the library documentation. Probably in the "providing an implementation" section. It could briefly explain how it's implemented too.
* I think it's fine for the std impl to require 1.63 and obviously therefore bump the MSRV, but I wonder if it would be kind to have a lower MSRV without the std feature, or at least mention it, as 1.63 is _so_ new and we're basically rolling this out across the whole embedded ecosystem. Since we'll probably add a CI test case for the feature, that one can build with the higher MSRV, and the other case could use the lower one. Dunno.

Dirbaio commented:
> (@eldruin) Could you also add this feature to the test CI workflow?

I didn't because it runs the doctests showing how to set an impl, which then fails duplicate symbol errors because there's also the std impl. IMO it should be fine since there are no tests other than the doctests anyway.

> But, it's more "obvious" to keep the lock until we've cleared the cell. What do you think?

Actually, we can also *set* the flag *before* acquiring the mutex. This way it's symmetric! :) 
I've added some comments highlighting the fact that it's fine due to being thread local.

> We should document the new feature in the README so that it also appears in the library documentation. Probably in the "providing an implementation" section. It could briefly explain how it's implemented too.

Done

> I think it's fine for the std impl to require 1.63 and obviously therefore bump the MSRV, but I wonder if it would be kind to have a lower MSRV without the std feature

SGTM. Documented MSRV as 1.54 without `std`, 1.63 with. Added CI checks for both. (1.54 needed for `#[doc(include_str!())]` support)

Dirbaio commented:
For some reason 1.54 doesn't like `#[transparent]` on `RestoreState`, so I've removed it.

```

error[E0690]: transparent struct needs exactly one non-zero-sized field, but has 0
   --> src/lib.rs:123:1
    |
123 | pub struct RestoreState(RawRestoreState);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ needs exactly one non-zero-sized field, but has 0
```

notgull commented:
I think it's because `RawRestoreState` can be a `()`, and ZSTs can't be the only field in a transparent struct.

Dirbaio commented:
Yeah, but that was fine in 1.63 :shrug:  Either way, I think *not* having it is fine, it's very unlikely it'll result in worse code. Ssee https://github.com/rust-embedded/critical-section/pull/20#discussion_r943389778

eldruin commented:
> > (@eldruin) Could you also add this feature to the test CI workflow?
> 
> I didn't because it runs the doctests showing how to set an impl, which then fails duplicate symbol errors because there's also the std impl. IMO it should be fine since there are no tests other than the doctests anyway.

Hmm, you can gate the doctest code with something like:
```rust
# #[cfg(not(feature = "std"))]
# {
// ...
# }
```

Dirbaio commented:
Done (it's slightly abominable, but works...)

adamgreig commented:
Thanks for the updates, LGTM.

eldruin reviewed (approved):
Alright, thank you everybody!
bors r+

eldruin commented:
Ach, @reitermarkus could you review this again?

Dirbaio commented:
bors r+

Dirbaio commented:
ugh, why does a "neutral" review with comments count as "rejected"?
